File: sdk/javascript/src/zkp.ts
/**
 * Zero-Knowledge Proofs (Schnorr)
 */

import { 
  randomBytes, sha256, bytesToHex, bytesToBigInt, bigIntToBytes 
} from './crypto';
import { ValidationError, AuthenticationError } from './errors';

// Parametres du groupe (RFC 5114)
const P = BigInt(
  '0xFFFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD1' +
  '29024E088A67CC74020BBEA63B139B22514A08798E3404DD' +
  'EF9519B3CD3A431B302B0A6DF25F14374FE1356D6D51C245' +
  'E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED' +
  'EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3D' +
  'C2007CB8A163BF0598DA48361C55D39A69163FA8FD24CF5F' +
  '83655D23DCA3AD961C62F356208552BB9ED529077096966D' +
  '670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B' +
  'E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9' +
  'DE2BCBF6955817183995497CEA956AE515D2261898FA0510' +
  '15728E5A8AACAA68FFFFFFFFFFFFFFFF'
);
const Q = (P - 1n) / 2n;
const G = 2n;

function modPow(base: bigint, exp: bigint, mod: bigint): bigint {
  let result = 1n;
  base = ((base % mod) + mod) % mod;
  while (exp > 0n) {
    if (exp % 2n === 1n) {
      result = (result * base) % mod;
    }
    exp = exp / 2n;
    base = (base * base) % mod;
  }
  return result;
}

export interface ZKPProof {
  commitment: string;  // R = g^k (hex)
  challenge: string;   // c (hex)
  response: string;    // s = k + c*x (hex)
  publicKey: string;   // Y = g^x (hex)
  message: string;     // hex
  timestamp: number;
}

export class ZKPProver {
  private readonly x: bigint;  // Cle privee
  private readonly Y: bigint;  // Cle publique
  
  constructor(vaultKey: Uint8Array) {
    if (vaultKey.length !== 32) {
      throw new ValidationError('vaultKey must be 32 bytes');
    }
    
    // Deriver cle privee
    const prefix = new TextEncoder().encode('PSNX_ZKP_PRIVATE_');
    const combined = new Uint8Array(prefix.length + vaultKey.length);
    combined.set(prefix);
    combined.set(vaultKey, prefix.length);
    
    // Note: synchrone sha256 via crypto.subtle pas dispo
    // On utilise une derivation simple pour l'instant
    let hash = 0n;
    for (let i = 0; i < combined.length; i++) {
      hash = (hash * 256n + BigInt(combined[i])) % Q;
    }
    
    this.x = hash === 0n ? 1n : hash;
    this.Y = modPow(G, this.x, P);
  }
  
  /**
   * Initialise avec hash async
   */
  static async create(vaultKey: Uint8Array): Promise<ZKPProver> {
    if (vaultKey.length !== 32) {
      throw new ValidationError('vaultKey must be 32 bytes');
    }
    
    const prefix = new TextEncoder().encode('PSNX_ZKP_PRIVATE_');
    const combined = new Uint8Array(prefix.length + vaultKey.length);
    combined.set(prefix);
    combined.set(vaultKey, prefix.length);
    
    const hash = await sha256(combined);
    let x = bytesToBigInt(hash) % Q;
    if (x === 0n) x = 1n;
    
    const prover = new ZKPProver(vaultKey);
    (prover as any).x = x;
    (prover as any).Y = modPow(G, x, P);
    
    return prover;
  }
  
  getPublicKey(): bigint {
    return this.Y;
  }
  
  getPublicKeyHex(): string {
    return '0x' + this.Y.toString(16);
  }
  
  async getFingerprint(): Promise<string> {
    const yBytes = bigIntToBytes(this.Y, 256);
    const hash = await sha256(yBytes);
    return bytesToHex(hash).slice(0, 16);
  }
  
  /**
   * Cree une preuve ZKP
   */
  async createProof(challenge: string): Promise<{
    proof: ZKPProof;
    challenge: string;
    keyFingerprint: string;
    timestamp: number;
  }> {
    const message = new TextEncoder().encode(challenge);
    const timestamp = Date.now() / 1000;
    
    // Commitment: R = g^k
    const kBytes = randomBytes(32);
    const k = (bytesToBigInt(kBytes) % (Q - 1n)) + 1n;
    const R = modPow(G, k, P);
    
    // Challenge: c = H(R || Y || m)
    const rBytes = bigIntToBytes(R, 256);
    const yBytes = bigIntToBytes(this.Y, 256);
    const data = new Uint8Array(rBytes.length + yBytes.length + message.length);
    data.set(rBytes);
    data.set(yBytes, rBytes.length);
    data.set(message, rBytes.length + yBytes.length);
    
    const cHash = await sha256(data);
    const c = bytesToBigInt(cHash) % Q;
    
    // Response: s = k + c*x mod Q
    let s = (k + c * this.x) % Q;
    if (s < 0n) s += Q;
    
    const proof: ZKPProof = {
      commitment: '0x' + R.toString(16),
      challenge: '0x' + c.toString(16),
      response: '0x' + s.toString(16),
      publicKey: '0x' + this.Y.toString(16),
      message: bytesToHex(message),
      timestamp
    };
    
    return {
      proof,
      challenge,
      keyFingerprint: await this.getFingerprint(),
      timestamp
    };
  }
}

export class ZKPVerifier {
  /**
   * Verifie une preuve ZKP
   */
  static async verify(
    authData: { proof: ZKPProof; challenge: string },
    expectedChallenge: string,
    maxAgeSeconds: number = 300
  ): Promise<{ valid: boolean; reason: string }> {
    try {
      const { proof } = authData;
      
      // Verifier challenge
      if (authData.challenge !== expectedChallenge) {
        return { valid: false, reason: 'Challenge mismatch' };
      }
      
      // Verifier age
      const age = Date.now() / 1000 - proof.timestamp;
      if (age > maxAgeSeconds) {
        return { valid: false, reason: `Proof expired (age: ${age.toFixed(1)}s)` };
      }
      if (age < -60) {
        return { valid: false, reason: 'Proof from future' };
      }
      
      // Parser les valeurs
      const R = BigInt(proof.commitment);
      const c = BigInt(proof.challenge);
      const s = BigInt(proof.response);
      const Y = BigInt(proof.publicKey);
      const message = new Uint8Array(
        proof.message.match(/.{2}/g)!.map(b => parseInt(b, 16))
      );
      
      // Verifier le challenge hash
      const rBytes = bigIntToBytes(R, 256);
      const yBytes = bigIntToBytes(Y, 256);
      const data = new Uint8Array(rBytes.length + yBytes.length + message.length);
      data.set(rBytes);
      data.set(yBytes, rBytes.length);
      data.set(message, rBytes.length + yBytes.length);
      
      const expectedC = bytesToBigInt(await sha256(data)) % Q;
      if (expectedC !== c) {
        return { valid: false, reason: 'Invalid challenge hash' };
      }
      
      // Verifier: g^s == R * Y^c mod p
      const lhs = modPow(G, s, P);
      const rhs = (R * modPow(Y, c, P)) % P;
      
      if (lhs !== rhs) {
        return { valid: false, reason: 'Cryptographic verification failed' };
      }
      
      return { valid: true, reason: 'OK' };
      
    } catch (e) {
      return { valid: false, reason: `Verification error: ${e}` };
    }
  }
  
  /**
   * Verifie une cle publique
   */
  static async verifyPublicKey(publicKey: bigint): Promise<boolean> {
    if (publicKey < 2n || publicKey >= P - 1n) {
      return false;
    }
    
    // Doit etre dans le sous-groupe
    return modPow(publicKey, Q, P) === 1n;
  }
}

File: sdk/python/psnx_sdk/zkp.py
"""
Zero-Knowledge Proofs pour le SDK Eidolon

Protocole de Schnorr pour prouver la possession d'une cle
sans la reveler.
"""

import secrets
import hashlib
import time
from dataclasses import dataclass
from typing import Dict, Tuple, Optional

from .exceptions import AuthenticationError, ValidationError


@dataclass
class ZKPProof:
    """Preuve Zero-Knowledge"""
    commitment: int  # R = g^k
    challenge: int   # c = H(R, Y, m)
    response: int    # s = k + c*x
    public_key: int  # Y = g^x
    message: bytes
    timestamp: float
    
    def to_dict(self) -> Dict:
        return {
            "commitment": hex(self.commitment),
            "challenge": hex(self.challenge),
            "response": hex(self.response),
            "public_key": hex(self.public_key),
            "message": self.message.hex(),
            "timestamp": self.timestamp
        }
    
    @classmethod
    def from_dict(cls, d: Dict) -> "ZKPProof":
        return cls(
            commitment=int(d["commitment"], 16),
            challenge=int(d["challenge"], 16),
            response=int(d["response"], 16),
            public_key=int(d["public_key"], 16),
            message=bytes.fromhex(d["message"]),
            timestamp=d["timestamp"]
        )


class ZKPParams:
    """Parametres du groupe pour Schnorr (RFC 5114)"""
    
    P = int(
        "FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD1"
        "29024E088A67CC74020BBEA63B139B22514A08798E3404DD"
        "EF9519B3CD3A431B302B0A6DF25F14374FE1356D6D51C245"
        "E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED"
        "EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3D"
        "C2007CB8A163BF0598DA48361C55D39A69163FA8FD24CF5F"
        "83655D23DCA3AD961C62F356208552BB9ED529077096966D"
        "670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B"
        "E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9"
        "DE2BCBF6955817183995497CEA956AE515D2261898FA0510"
        "15728E5A8AACAA68FFFFFFFFFFFFFFFF", 16
    )
    Q = (P - 1) // 2
    G = 2


class ZKPProver:
    """
    Prover pour Zero-Knowledge Proofs (cote client).
    
    Usage:
        prover = ZKPProver(vault_key)
        
        # Obtenir la cle publique
        public_key = prover.get_public_key()
        
        # Creer une preuve
        proof = prover.create_proof(challenge)
    """
    
    def __init__(self, vault_key: bytes):
        """
        Args:
            vault_key: Cle du vault (32 bytes)
        """
        if len(vault_key) != 32:
            raise ValidationError("vault_key must be 32 bytes")
        
        # Deriver la cle privee
        h = hashlib.sha256(b"PSNX_ZKP_PRIVATE_" + vault_key).digest()
        self._x = int.from_bytes(h, 'big') % ZKPParams.Q
        if self._x == 0:
            self._x = 1
        
        # Calculer la cle publique
        self._Y = pow(ZKPParams.G, self._x, ZKPParams.P)
        
        # Commitment temporaire
        self._k: Optional[int] = None
    
    def get_public_key(self) -> int:
        """Retourne la cle publique"""
        return self._Y
    
    def get_public_key_hex(self) -> str:
        """Retourne la cle publique en hex"""
        return hex(self._Y)
    
    def get_fingerprint(self) -> str:
        """Retourne l'empreinte"""
        return hashlib.sha256(
            self._Y.to_bytes((self._Y.bit_length() + 7) // 8, 'big')
        ).hexdigest()[:16]
    
    def create_proof(self, challenge: str) -> Dict:
        """
        Cree une preuve ZKP pour un challenge.
        
        Args:
            challenge: Challenge du serveur
        
        Returns:
            Preuve en format dict (pour API)
        """
        message = challenge.encode()
        
        # Commitment: R = g^k
        self._k = secrets.randbelow(ZKPParams.Q - 1) + 1
        R = pow(ZKPParams.G, self._k, ZKPParams.P)
        
        # Challenge: c = H(R || Y || m)
        c = self._compute_challenge(R, self._Y, message)
        
        # Response: s = k + c*x mod Q
        s = (self._k + c * self._x) % ZKPParams.Q
        
        # Effacer k
        self._k = None
        
        proof = ZKPProof(
            commitment=R,
            challenge=c,
            response=s,
            public_key=self._Y,
            message=message,
            timestamp=time.time()
        )
        
        return {
            "proof": proof.to_dict(),
            "challenge": challenge,
            "key_fingerprint": self.get_fingerprint(),
            "timestamp": proof.timestamp
        }
    
    def _compute_challenge(self, R: int, Y: int, message: bytes) -> int:
        """Calcule le challenge (Fiat-Shamir)"""
        r_bytes = R.to_bytes((R.bit_length() + 7) // 8, 'big')
        y_bytes = Y.to_bytes((Y.bit_length() + 7) // 8, 'big')
        
        data = r_bytes + y_bytes + message
        h = hashlib.sha256(data).digest()
        
        return int.from_bytes(h, 'big') % ZKPParams.Q


class ZKPVerifier:
    """
    Verifier pour Zero-Knowledge Proofs (cote serveur).
    
    Usage:
        # Verification d'une preuve
        valid = ZKPVerifier.verify(proof_dict, expected_challenge)
    """
    
    @staticmethod
    def verify(
        auth_data: Dict,
        expected_challenge: str,
        max_age_seconds: float = 300
    ) -> Tuple[bool, str]:
        """
        Verifie une preuve ZKP.
        
        Args:
            auth_data: Donnees d'authentification
            expected_challenge: Challenge attendu
            max_age_seconds: Age maximum de la preuve
        
        Returns:
            (valid, reason)
        """
        try:
            # Extraire la preuve
            proof = ZKPProof.from_dict(auth_data["proof"])
            
            # Verifier le challenge
            if auth_data.get("challenge") != expected_challenge:
                return False, "Challenge mismatch"
            
            # Verifier l'age
            age = time.time() - proof.timestamp
            if age > max_age_seconds:
                return False, f"Proof expired (age: {age:.1f}s)"
            if age < -60:
                return False, "Proof from future"
            
            # Recalculer le challenge attendu
            r_bytes = proof.commitment.to_bytes(
                (proof.commitment.bit_length() + 7) // 8, 'big'
            )
            y_bytes = proof.public_key.to_bytes(
                (proof.public_key.bit_length() + 7) // 8, 'big'
            )
            data = r_bytes + y_bytes + proof.message
            expected_c = int.from_bytes(
                hashlib.sha256(data).digest(), 'big'
            ) % ZKPParams.Q
            
            if expected_c != proof.challenge:
                return False, "Invalid challenge hash"
            
            # Verifier l'equation: g^s == R * Y^c mod p
            lhs = pow(ZKPParams.G, proof.response, ZKPParams.P)
            rhs = (
                proof.commitment *
                pow(proof.public_key, proof.challenge, ZKPParams.P)
            ) % ZKPParams.P
            
            if lhs != rhs:
                return False, "Cryptographic verification failed"
            
            return True, "OK"
            
        except Exception as e:
            return False, f"Verification error: {e}"
    
    @staticmethod
    def verify_public_key(public_key: int) -> bool:
        """Verifie qu'une cle publique est valide"""
        # Doit etre dans [2, P-2]
        if public_key < 2 or public_key >= ZKPParams.P - 1:
            return False
        
        # Doit etre dans le sous-groupe (Y^Q == 1 mod P)
        if pow(public_key, ZKPParams.Q, ZKPParams.P) != 1:
            return False
        
        return True
